Create own codegen

koriym opened the pull request:
* Remove nikic/php-parser
* Motivated by increasing speed and reducing dependence
* Not easy job

## 独自のcodegen実装が困難な理由

nikic/php-parserを使わないで独自のcodegenでオリジナルのコードを変更するのは困難です。以下がその理由。

* リフレクションだけでは必要な情報が集められない。例えばuseが取得できない。
* token_get_allで得られるtokenでは内容の改変が困難。
* 親クラスのクラスのメソッドも取得して１つののコードにする必要あるが、親クラスの必要な情報だけをtokenからだけ取得するのは非常に困難。

## 解決策

*　token_get_allで得られるtokenとリフレクションのハイブリッドで行う
*　classの宣言まではtokenとその改変で行う
*　メソッド情報はリフレクションから取得。これは引数の名前解決がtokenでは困難なため。 

そもそもこの実装が最適化されたとしてもどの程度パフォーマンスアップ、低メモリになるかは未知数ですが実装してみないことにはわかりません。いうまでもありませんがプロダクションではコンパイルされたコードが使われるのでAOPによるパフォーマンスインパクトはほとんどありません。このPRは開発時やディプロイの時の最初のヘルスチェックでコンパイルを行う時のパフォーマンス向上です。

## Not easy

It is difficult to modify the original code with our own codegen without using nikic/php-parser. Here is why.

* Reflection alone cannot gather necessary information. For example, `use` cannot be obtained.
* It is difficult to modify the contents with TOKEN obtained by token_get_all.
* It is very difficult to get only necessary information of the parent class from token, although it is necessary to get methods of the parent class to make a single code.

## Solution

* Use a hybrid of token and reflection obtained by token_get_all.
* Use token and its modification to declare classes.
* Method information is obtained from reflection. This is because name resolution of arguments is difficult with token. 

Even if this implementation is optimized, it is unknown how much performance and memory reduction will be achieved. Needless to say, the performance impact of AOP will be negligible in production, where compiled code is used. The PR is a performance improvement when compiling at the first health check during development and deployment.

## Todo

 * [x] unionタイプのリターンテスト
 * [x] 導入する価値があるかベンチマーク x20+
 * [x] SA & coverage 100%
 * [x] アトリビュートの変換を名前付き引数に
 
 Related: https://github.com/ray-di/Ray.Aop/issues/206
 
Closes #206 

----

[パーサトークンの一覧](https://www.php.net/manual/ja/tokens.php#tokens)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Introduced new classes for aspect-oriented programming enhancements.
	- Added support for PHP 8.2.
	- Enhanced functionality in `ReflectionClass` and other components to improve code handling and type management.

- **Bug Fixes**
	- Adjusted logic and fixed issues in various classes to ensure smoother operation and interaction.

- **Refactor**
	- Refactored several components to optimize performance and maintainability.

- **Tests**
	- Added and updated tests to cover new functionality and ensure reliability.

- **Documentation**
	- Updated configuration files to align with new features and standards.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

koriym committed:
Add AI-generated codegen

https://chat.openai.com/share/eb4adfc3-c836-4ca8-9f77-4245bfa3ba96

koriym committed:
Create AopCodeGen

Create aop class without php parser

koriym committed:
Remove unbound method

koriym committed:
Create AopCodeGenNewCode

koriym committed:
Do not add properties and constants to AOP class

koriym committed:
Add parent method

but not implemented for parent "use" namespace.

koriym committed:
Get method signature of parents class by reflection

The parent class also called generate to get the method string in the same way, but type resolution of the arguments proved difficult and was done by reflection.

koriym committed:
implements WeaveInterface

koriym committed:
Generate methods from reflection

100% test passed in this point

koriym committed:
Remove unused token based code

koriym committed:
Move nikic/php-parser to require-dev

Still old test code exists

koriym committed:
Move nikic/php-parser to require-dev

Still old test code exists

koriym committed:
Fix regex

Thanks phpstorm notice

koriym committed:
Revert "Add AI-generated codegen"

This reverts commit 99a28a757d8f4d94b9407f90a693482dbec51c94.

koriym committed:
Create union type test and fix

koriym committed:
Extract AopCodeGen method signature

koriym committed:
Soothe psalm

koriym committed:
Soothe phpstan

koriym committed:
No more support MultipleClassInOneFileException

It's hard for this codegen.

koriym committed:
Bump QA tools

koriym committed:
Test coverage 100%

koriym committed:
Soothe CS

koriym committed:
Enable PHP 7.2 compat

koriym committed:
Soothe phpmd

koriym committed:
Soothe composer-require-checker

koriym committed:
fixup! Generate methods from reflection

koriym committed:
Do not check PHP_VERSION on runtime

koriym committed:
Move php-parser based codegen deprecated test

koriym committed:
CI with PHP 8.2

koriym committed:
Ignore .phpunit.cache

koriym committed:
Refactor InterceptTrait

koriym committed:
fixup! Refactor InterceptTrait

Missed because PHPStorm detected false positive

koriym committed:
No root level sign "\" for self

koriym committed:
Suppress CyclomaticComplexity

koriym commented:
@todo アトリビュートの変換を名前付き引数に

~今は引数に名前がついていないので、doctrineアノテーションで古いコンストラクタの使用（arrayで受け取る）事を考慮したコードだと順番引数と相違が出てエラーになってしまいます。~

解決しました。

koriym committed:
Support for cases where attributes use named parameters.

koriym committed:
Enable PHP 7.2 compat

koriym committed:
Remove src-deprecated from autoload

jingu commented on src/AopCodeGenMethodSignature.php:
```diff
@@ -0,0 +1,138 @@
+
+        if ($type instanceof ReflectionNamedType) {
+            /** @psalm-suppress TypeDoesNotContainType */
+            $typeStr = $type->isBuiltin() || $type->getName() === 'self' ? $type->getName() : '\\' . $type->getName();
```
```suggestion
            $typeStr = $type->isBuiltin() || $type->getName() === 'self' || $type->getName() === 'static' ? $type->getName() : '\\' . $type->getName();
```

jingu commented on src/AopCodeGenNewCode.php:
```diff
@@ -0,0 +1,67 @@
+
+    public function implementsInterface(string $interfaceName): void
+    {
+        $pattern = '/(class\s+\w+\s+(?:extends\s+\w+\s+)?)(implements\s+\w+(?:,\s*\w+)*)?/';
```
```suggestion
        $pattern = '@(class\s+\w+\s*(?:extends\s+\w+)?)\s*(implements\s+[\w]+)@';
```

koriym committed:
Update src/AopCodeGenNewCode.php

Co-authored-by: Yoshitaka Jingu <yoshitaka.jingu@gmail.com>

koriym committed:
Update src/AopCodeGenMethodSignature.php

Co-authored-by: Yoshitaka Jingu <yoshitaka.jingu@gmail.com>

koriym committed:
Throw CompilationFailedException when compilation failed

koriym committed:
Test FQN interface

koriym committed:
Add space between extends and implementation part

koriym committed:
Update regex to match backslashed namespace

koriym committed:
Refactor: Remove elseif

koriym committed:
Refactor: EarlyExit

koriym committed:
fixup! Refactor: Remove elseif

koriym committed:
fixup! Refactor: EarlyExit

koriym committed:
Coverage 100%

koriym committed:
[ci skip] Update composer command path

koriym committed:
fixup! fixup! Refactor: EarlyExit

NaokiTsuchiya commented on src-deprecated/README.md:
````diff
@@ -0,0 +1,7 @@
+
+```json
+"psr-4": {
+            "Ray\\Aop\\": ["vendor/reay/aop/src-deprecated"]
````
```suggestion
            "Ray\\Aop\\": ["vendor/ray/aop/src-deprecated"]
```

NaokiTsuchiya replied:
require から php-parser が外れたので、 php-parser のインストールが必要であることもガイドした方が良いですかね？

koriym replied:
そもそも、src-deprecatedを丸ごと削除しても？

NaokiTsuchiya commented on src-deprecated/CallIntercept.php:
```diff
@@ -10,7 +10,8 @@
 use PhpParser\NodeAbstract;
 
-final class CallIntercept
+final class
```
```suggestion
/** @deprecated */
final class
```

koriym replied:
利用も想定されないのでsrc-deprecatedを削除しました。

NaokiTsuchiya commented on src-deprecated/CallIntercept.php:
```diff
@@ -37,7 +38,7 @@ public function __construct(BuilderFactory $factory)
     }
 
-    /** @return list<Expression>|list<Return_> */
+    /** @return Return_|Return_ */
```
ここは変更必要なんですかね？

実装が変わっていないので変わらないと思うのですが。。

NaokiTsuchiya commented on tests/Fake/FakePhp8Types.php:
```diff
@@ -0,0 +1,65 @@
+use Ray\Aop\Annotation\FakeMarker4;
+use Ray\Aop\Annotation\FakeMarkerName;
+
+class FakePhp8Types implements FakeNullInterface, \Ray\Aop\FakeNullInterface1
```
8.2でサポートされた DNF 型 や `false`, `true`, `null` などのケースもあった方がよいですかね？

NaokiTsuchiya replied:
以下のケースはパスしました。

* `public function method23(): false { return false; }`
* `public function method24(): true { return true; }`

以下のケースは失敗しました。

* `public function method25(): FakeNullInterface&FakeNullInterface1 { return $this;}`
* `public function method27(): null { return null; }`

`public function method25(): FakeNullInterface&FakeNullInterface1 { return $this;}`

```
assert($type instanceof ReflectionNamedType) in /Users/tsuchiya/work/github.com/ray-di/Ray.Aop/src/AopCodeGenMethodSignature.php:135
 
 Caused by
 AssertionError: assert($type instanceof ReflectionNamedType)
 
 /Users/tsuchiya/work/github.com/ray-di/Ray.Aop/src/AopCodeGenMethodSignature.php:135
 /Users/tsuchiya/work/github.com/ray-di/Ray.Aop/src/AopCodeGenMethodSignature.php:87
 /Users/tsuchiya/work/github.com/ray-di/Ray.Aop/src/AopCodeGen.php:113
 /Users/tsuchiya/work/github.com/ray-di/Ray.Aop/src/AopCodeGen.php:96
 /Users/tsuchiya/work/github.com/ray-di/Ray.Aop/tests/AopCodeGenTest.php:54
```

`public function method27(): null { return null; }`

```
PHP Fatal error:  Duplicate type null is redundant in /private/var/folders/vg/stmrlsw543z4k90c37v9xq9w0000gq/T/tmp_BLGNl7.php on line 143

Fatal error: Duplicate type null is redundant in /private/var/folders/vg/stmrlsw543z4k90c37v9xq9w0000gq/T/tmp_BLGNl7.php on line 143

Process finished with exit code 255
```

koriym committed:
Apply suggestions from code review

Co-authored-by: Naoki Tsuchiya <tsuchiya@bengo4.com>

koriym committed:
Remove src-deprecated

koriym committed:
Rename variable

koriym committed:
Rename test method name

koriym committed:
Skip constuctor method for binding

koriym committed:
Support PHP 8.2 DNF type

koriym committed:
Extract TypeString

koriym committed:
Rename gen code classes

koriym committed:
Set php target 8.2 to analyze code

koriym committed:
fixup! Skip constuctor method for binding

koriym committed:
Rquire php 8.2 for DNF type test

koriym committed:
Move addMethods to GeneratedCode

koriym committed:
fixup! Set php target 8.2 to analyze code

koriym committed:
Refactor AopCodeGen

koriym committed:
Create TokenIterator

koriym committed:
Move parseClass method

koriym committed:
Create AopCodeGen::generate method

koriym committed:
Rename GeneratedCode to AopCode

koriym committed:
Compromise CyclomaticComplexity in phpmd
from 10 to 11

koriym committed:
Remove AopCodeGen

koriym committed:
fixup! Remove AopCodeGen

koriym commented:
@NaokiTsuchiya PHP8.2のDNF型、その他に対応しました。
https://github.com/ray-di/Ray.Aop/blob/8507909b4f92148026c795329f89c4c724a675a0/tests/Fake/FakePhp82Types.php

koriym commented:
リファクタリングしてAopCodeGenを廃止し、AopCode(GeneratedCode)に集約しました。(OOP)

https://github.com/ray-di/Ray.Aop/pull/205/files#diff-27283633b526d5b90eca1c0d03ff97fd779708eb2ecae60ece12bae1775ae110R47-R49

koriym committed:
fixup! Support PHP 8.2 DNF type

jingu commented on src/TypeString.php:
```diff
@@ -0,0 +1,87 @@
+
+    private static function getFqnType(ReflectionNamedType $type): string
+    {
+        $isBuiltinOrSelf = $type->isBuiltin();
```
```suggestion
        $isBuiltin = $type->isBuiltin() || in_array($type->getName(), ['static', 'self'], true);
```

jingu commented on src/TypeString.php:
```diff
@@ -0,0 +1,87 @@
+    {
+        $isBuiltinOrSelf = $type->isBuiltin();
+
+        return $isBuiltinOrSelf ? $type->getName() : '\\' . $type->getName();
```
```suggestion
        return $isBuiltin ? $type->getName() : '\\' . $type->getName();
```

jingu reviewed (commented):
手元のプロジェクトではメソッドの戻り値がstaticでエラーが出ましたのでsuggestしました。

koriym committed:
Update src/TypeString.php

Co-authored-by: Yoshitaka Jingu <yoshitaka.jingu@gmail.com>

koriym committed:
Update src/TypeString.php

Co-authored-by: Yoshitaka Jingu <yoshitaka.jingu@gmail.com>

jingu reviewed (approved):
👍

koriym committed:
fixup! Update src/TypeString.php

koriym committed:
Soothe sa

koriym committed:
Update SA tool and soothe them

The commit introduces `is_string` function in a condition in `MethodSignatureString.php`. Several tools in `composer.lock` have been updated to newer versions, including packages like `amphp/amp`, `amphp/byte-stream`, and others. This commit thus refactors the code for improved validation and updates dependencies for enhanced project features.

koriym commented:
@coderabbitai 日本語でもう一度お願いでします。

koriym commented:
@coderabbitai Recent Review Detailsの部分も日本語でお願いします。

koriym commented:
@NaokiTsuchiya mergeしました！
